### PR TITLE
[hailctl] fix hailctl batch submit and restore tests

### PR DESCRIPTION
**Pull request by danking**
Fixes: #14247

Issues resolved herein:

1. build.yaml tests must not use `exit 0` as it exits the test early.
2. Always prefer `orjson` to `json`.
3. Add `--wait` which waits for the submitted batch to complete and exits success only when the batch is success.
4. Whenever working with paths, we must use the `realpath` which resolves symlinks. In particular, on Mac OS X, `/tmp` is a symlink to `/private/tmp` and Python's APIs are inconsistent on whether they return a realpath or a path with symlinks. [1]
5. If the destination looks like a directory (e.g. "bar:/foo/", "bar:/"), the tests all suggest we should copy *into* not *to*. We now check for a trailing slash and copy *into*.
6. `ln -s src dst` means different things depending on whether dst is an extant folder or not. In this PR, I prefer to always be fully explicit so I never rely on `ln` detecting the destination is a directory and acting differently. Put differently: `file_input_to_src_dest` now never returns a file source and a destination folder.
7. We need to create the `real_absolute_cwd()` on the job before we `cd` into it.
8. `test_dir_outside_curdir` suggests that `--file foo/:/` is meant to copy the contents of foo into the root. This cannot be implemented with our symlink strategy (you can't replace the root with a symlink), so I changed the interpretation: a trailing slash on the source is meaningless. If the destination ends in a slash, we "copy into", otherwise we "copy to".
9. Add examples of --files usage.

[1]:

```ipython3
In [1]: import tempfile
   ...: tempfile.TemporaryDirectory()
Out[1]: <TemporaryDirectory '/var/folders/x1/601098gx0v11qjx2l_7qfw2c0000gq/T/tmp_pmj3lr9'>

In [2]: import os
   ...: os.getcwd()
Out[2]: '/private/tmp'

In [3]: !ls -al /tmp
lrwxr-xr-x@ 1 root  wheel  11 Aug  2 05:44 /tmp -> private/tmp
```

### CODE REVIEW

File: hail/python/hailtop/hailctl/batch/submit.py [hunk 1]
```diff
@@ -46,14 +64,17 @@ def file_input_to_src_dest(file: str) -> Tuple[str, str, str]:
...
         dest = result.get('dest')
         if dest is not None:
-            dest = os.path.abspath(os.path.expanduser(dest))
+            dest_intended_as_directory = dest[-1] == '/'
+            dest = real_absolute_expanded_path(dest)
+            if dest_intended_as_directory:
+                dest = os.path.join(dest, os.path.basename(src))
         else:
-            dest = os.getcwd()
+            dest = os.path.join(real_absolute_cwd(), os.path.basename(src))
```
Comment (danking) [inline hail/python/hailtop/hailctl/batch/submit.py:77]: @jigold commented at https://github.com/hail-is/hail/pull/14183:

> Can you explain this change?

The semantics of `ln -s src dst` depends on whether `dst` is a directory or not. I found it easier to reason about `file_input_to_src_dest` when I had the post-condition: for each triplet, the destination is the desired location of a symlink to the source.

Before this change the post-condition was "the destination is the desired location of a symlink to the source or a folder into which we want a symlink placed". The latter situation feels like a footgun. The presence or absence of a trailing slash (which realname removes) affects the semantics. Moreover `basename /foo/` prints `foo` whereas `os.path.basename('/foo/')` returns `''`.

With these changes, we still use the trailing slash to determine if we want copy-to or copy-into, but once that's resolved, we're always dealing with fully specified source-destination file-pairs.

**Comment by daniel-goldstein** [comment]
See https://hail.zulipchat.com/#narrow/stream/127527-team/topic/cloud.20fuse.20failing.20tests/near/417695786 for why I added do-not-test

File: build.yaml [hunk 2]
```diff
@@ -3093,45 +3097,39 @@ steps:
...
-      if [ "$STATE" == "success" ]; then
-          exit 0;
-      else
+      if [ "$STATE" != "success" ]; then
           echo $STATUS;
           exit 1;
       fi
 
-      cat >file.sh <<EOF
+      cat >file.sh <<'EOF'
```
Comment (jigold) [inline build.yaml:3105]: Is this a typo?

File: build.yaml [hunk 3]
```diff
@@ -3093,45 +3097,39 @@ steps:
...
-      if [ "$STATE" == "success" ]; then
-          exit 0;
-      else
+      if [ "$STATE" != "success" ]; then
           echo $STATUS;
           exit 1;
       fi
 
-      cat >file-with-args.sh <<EOF
+      cat >file-with-args.sh <<'EOF'
```
Comment (jigold) [inline build.yaml:3120]: same question about typo in `'EOF'`.

File: hail/python/test/hailtop/hailctl/batch/test_submit.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+@pytest.mark.timeout(5 * 60)  # image pulling is very slow
+def test_dir_outside_curdir(runner: CliRunner):
+    with tempfile.TemporaryDirectory() as dir:
+        os.mkdir(f'{dir}/working_dir')
+        os.chdir(f'{dir}/working_dir')
+        write_hello(f'{dir}/hello1.txt')
+        write_hello(f'{dir}/hello2.txt')
+
+        dir_basename = os.path.basename(dir)
+        write_script(dir, f'/{dir_basename}/hello1.txt')
```
Comment (jigold) [inline hail/python/test/hailtop/hailctl/batch/test_submit.py:135]: Can you explain why the behavior of needing to use `basename` is different than in `test_dir_mount_in_child_dir_to_child_dir`?

**Review by jigold — CHANGES_REQUESTED** [review]

File: build.yaml (hunk 2 above)
Comment (danking) [reply build.yaml:3105]: No, it prevents interpolation of variables inside the heredoc. From the bash man page (key part is "If any characters in word are quoted, [...] the lines in the here-document are not expanded"):
> Here Documents
>     This type of redirection instructs the shell to read input from the current source until a line containing only word (with no trailing
>     blanks) is seen.  All of the lines read up to that point are then used as the standard input for a command.
>
>     The format of here-documents is:
>
>            <<[-]word
>                    here-document
>            delimiter
>
>     No parameter expansion, command substitution, arithmetic expansion, or pathname expansion is performed on word.  If any characters in word
>     are quoted, the delimiter is the result of quote removal on word, and the lines in the here-document are not expanded.  If word is unquoted,
>     all lines of the here-document are subjected to parameter expansion, command substitution, and arithmetic expansion.  In the latter case,
>     the character sequence \<newline> is ignored, and \ must be used to quote the characters \, $, and `.
>
>     If the redirection operator is <<-, then all leading tab characters are stripped from input lines and the line containing delimiter.  This
>     allows here-documents within shell scripts to be indented in a natural fashion.

File: hail/python/test/hailtop/hailctl/batch/test_submit.py (hunk 4 above)
Comment (danking) [reply hail/python/test/hailtop/hailctl/batch/test_submit.py:135]: Yeah I think maybe there's still a bug here. The semantics of `/` that hailctl currently uses doesn't work well with symlinks.

In `test_dir_mount_in_child_dir_to_child_dir`:
```
--files child/:/child/
```
The trailing slash on the source, `child/`, indicates we want to work with the contents of child. The trailing slash on the destination, `/child/`, indicates we want the destination to be *inside* `/child/`. This means hello1.txt and hello2.txt are placed inside the folder `/child/`. To be clear: I deduced these rules from the code, I'm not married to these being the rules.

In contrast, in `test_dir_outside_curdir`:
```
--files {dir}/:/
```
This *should* have the same semantics as above but we cannot symlink to `/` (the OS already exists there) so the only option is to put it in `/the-basename`. I suspect the code just breaks currently. We should maybe prohibit `x/:/`.

Comment (jigold) [reply hail/python/test/hailtop/hailctl/batch/test_submit.py:135]: I think we should prohibit it if we're doing something special.

**Comment by jigold** [comment]
Closing in favor of #14351

**Pull request closed without merging**